### PR TITLE
[GPT-3]add support for pure fp16 in dygraph mode

### PR DESCRIPTION
**Pull request by haohongxiang**
### PR types

New features
### PR changes

Others
### Description

[GPT-3]add support for pure fp16 in dygraph mode

单卡FP32与3D混合并行+recompute+FP16的loss曲线精度对比
image

### CODE REVIEW

File: examples/language_model/gpt-3/dygraph/args.py [hunk 1]
```diff
@@ -220,13 +220,13 @@ def parse_args(MODEL_CLASSES):
         const=False,
         help="Using the recompute to save the memory.")
 
-    # AMP config
+    # Pure FP16 config
     parser.add_argument(
-        "--use_amp",
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/args.py:225]: 是否以后就只用 `prue_fp16` 了，不再暴露 `amp` 选项给用户？

File: examples/language_model/gpt-3/dygraph/run.sh [hunk 2]
```diff
@@ -1,3 +1,5 @@
+export PYTHONPATH=$PYTHONPATH:../../../../
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run.sh]: 合入的时候，删去此行。默认用户需要安装paddlenlp

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 3]
```diff
@@ -246,16 +246,22 @@ def do_train(args):
             parameters=model.parameters(),
             weight_decay=args.weight_decay,
             grad_clip=clip,
-            apply_decay_param_fun=lambda x: x in decay_params)
+            apply_decay_param_fun=lambda x: x in decay_params,
+            multi_precision=args.use_pure_fp16)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:253]: 从API使用建议的角度来讲：
既然我们的用法是`paddle.amp.decorate`去包装optimizer。
这里的 `multi_precision` 选项的操作，如果没有更多使用场景需要 解耦的话，
放到`paddle.amp.decorate`中去操作，是更为合理的。

只是建议，可忽略。cc @sandyhouse

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 4]
```diff
@@ -246,16 +246,22 @@ def do_train(args):
...
+            level='O2',
+            save_dtype='float32')
 
     if paddle.distributed.get_world_size() > 1:
         model = fleet.distributed_model(model)
         optimizer = fleet.distributed_optimizer(optimizer)
 
-    if args.use_amp:
-        scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
-        scaler = fleet.distributed_scaler(scaler)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:257]: 所以普通的amp的话不用`paddle.amp.decorate`去装饰，这个是amp和pure fp16的主要diff？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 5]
```diff
@@ -246,16 +246,22 @@ def do_train(args):
...
             weight_decay=args.weight_decay,
             grad_clip=clip,
-            apply_decay_param_fun=lambda x: x in decay_params)
+            apply_decay_param_fun=lambda x: x in decay_params,
+            multi_precision=args.use_pure_fp16)
+
+    if args.use_pure_fp16:
+        scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
+        scaler = fleet.distributed_scaler(scaler)
+        model, optimizer = paddle.amp.decorate(
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:259]: 可以加一行注释，说明，`level O2 means to convert the network to FP16`

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 6]
```diff
@@ -294,21 +300,31 @@ def do_train(args):
...
-                                "c_embedding"
-                            ]):
-                        preds = model(tokens, position_ids)
-                        loss = criterion(preds, labels, loss_mask)
-
-                    if args.use_amp:
+                    lbs = args.local_batch_size
+                    mbs = args.micro_batch_size
+                    loss = 0.0
+                    for i in range((int)(lbs / mbs)):
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 额，不是太理解了？这里是为了做gradient merge？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 7]
```diff
@@ -294,21 +300,31 @@ def do_train(args):
...
+                                preds, labels[i * mbs:i * mbs + mbs, :],
+                                loss_mask[i * mbs:i * mbs + mbs, :])
+                        loss += loss_mbs
+                    loss = loss / (int)(lbs / mbs)
+
+                    if args.use_pure_fp16:
                         scaler.scale(loss).backward()
-                        scaler.minimize(optimizer, loss)
+                        scaler.step(optimizer)
+                        scaler.update()
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: `scaler.minimize(optimizer, loss)`
-> `scaler.step(optimizer), scaler.update()` 

这种用法的变化，可否统一？

**Review by ZHUI — CHANGES_REQUESTED** [review]
总体来看：
 amp和pure_fp16的一些API使用上，存在一些diff，建议统一。
cc @sandyhouse  @ForFishes 

如果需要同时支持 amp 和 pure_fp16 的,  参数选项可以加 一个 fp16_level, 默认 O2， `use_amp`改成`use_fp16`

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 3 above)
Comment (zhiqiu) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:253]: > 从API使用建议的角度来讲： 既然我们的用法是`paddle.amp.decorate`去包装optimizer。 这里的 `multi_precision` 选项的操作，如果没有更多使用场景需要 解耦的话， 放到`paddle.amp.decorate`中去操作，是更为合理的。
> 
> 只是建议，可忽略。cc @sandyhouse

嗯嗯，这是我们设计接口decorate的初衷，也是不希望在声明optimizer中有耦合amp的东西。

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 4 above)
Comment (zhiqiu) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:257]: > 所以普通的amp的话不用`paddle.amp.decorate`去装饰，这个是amp和pure fp16的主要diff？

amp O1可以用decorate（会不作操作直接返回），也可以不用。

File: examples/language_model/gpt-3/dygraph/args.py (hunk 1 above)
Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/args.py:225]: 是的，以后只用pure_fp16，不再暴露amp

File: examples/language_model/gpt-3/dygraph/run.sh (hunk 2 above)
Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/run.sh]: 好的

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 8]
```diff
@@ -294,21 +300,31 @@ def do_train(args):
                 position_ids.stop_gradient = True
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:304]: 再请教一个问题。

如果模型的输入是fp32的话， pure_fp16 需不需要对输入dataset的输入也转换为 fp16，即在数据处理阶段，将数据直接转为fp16，减少训练时候cast的开销？

（gpt的话大部分数据tokens是int类型的）

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 3 above)
Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:253]: 嗯嗯，了解了哈👌🏻

现阶段可以去掉 `multi_precision `的话，就直接去掉，不行的话，咱们记一个todo吧 @haohongxiang

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 4 above)
Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:257]: 普通的`amp`也用`paddle.amp.decorate`去装饰，用参数`level`区分，'O1'为`amp`，'O2'为`pure fp16`

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 3 above)
Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:253]: 好的

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 5 above)
Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:259]: > 可以加一行注释，说明，`level O2 means to convert the network to FP16`

好的

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 6 above)
Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: > 额，不是太理解了？这里是为了做gradient merge？

这里是在非pp情况下做各micro batch size的梯度累加

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 7 above)
Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: > `scaler.minimize(optimizer, loss)` -> `scaler.step(optimizer), scaler.update()`
> 
> 这种用法的变化，可否统一？

好的，可以统一

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 6 above)
Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 给一点建议
1. accumulate_steps 的计算，放到和 strategy.pipeline_configs 代码附近，统一计算accumulate_steps
2. 这里的 (int)(lbs / mbs) 理论上是需要check是否严格倍数关系，否则需要报错
3. 我看你的 这个累加阶段，只有forward，backward 在for循环外，是否有bug？
3. 这里的 gradient merge 是否需要 no_sync 提升性能 https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/DataParallel_cn.html#no_sync

Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 小建议:
gradient merge 策略加入到模型精度的CI  cc @ForFishes

Comment (haohongxiang) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: > 给一点建议
> 
> 1. accumulate_steps 的计算，放到和 strategy.pipeline_configs 代码附近，统一计算accumulate_steps
> 2. 这里的 (int)(lbs / mbs) 理论上是需要check是否严格倍数关系，否则需要报错
> 3. 我看你的 这个累加阶段，只有forward，backward 在for循环外，是否有bug？
> 4. 这里的 gradient merge 是否需要 no_sync 提升性能 https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/DataParallel_cn.html#no_sync

这边的代码有点问题，我修改一下

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 9]
```diff
@@ -320,19 +343,17 @@ def do_train(args):
                 else:
                     data = [(tokens, position_ids), (labels, loss_mask)]
                     with paddle.amp.auto_cast(
-                            args.use_amp,
-                            custom_white_list=[
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:324]: 这里的 `custom_white_list` 和 `custom_black_list` 修改是否有影响?

**Review by ZHUI — APPROVED** [review]
LGTM
